File: src/1115/main.go
// class FooBar {
// 	public void foo() {
//       for (int i = 0; i < n; i++) {
//         print("foo");
//       }
// 	}

// 	public void bar() {
//       for (int i = 0; i < n; i++) {
//         print("bar");
//       }
// 	}
//   }
//   两个不同的线程将会共用一个 FooBar 实例。其中一个线程将会调用 foo() 方法，另一个线程将会调用 bar() 方法。

//   请设计修改程序，以确保 "foobar" 被输出 n 次。

package main

import (
	//"fmt"
	"fmt"
	"sync"
)

type FooBar struct{
	n int 
}


func(f * FooBar) foo(){
	for i:=0;i<f.n;i++{
		fmt.Print("foo")
	}
}

func(f * FooBar) bar(){
	for i:=0;i<f.n;i++{
		fmt.Print("bar")
	}
}

func main() {
	defer func(){
		if errs:=recover();errs!=nil{
			fmt.Println(errs)
		}
	}()

	f := new(FooBar)

	f.n = 10
	
	wg := sync.WaitGroup{}

	wg.Add(f.n)

	Lock := sync.Mutex{}

	for i:=0 ;i<f.n;i++{
		go func(f * FooBar){
			Lock.Lock()
			f.foo()
			f.bar()
			fmt.Print("\n")
			defer Lock.Unlock()
			wg.Done()
		}(f)
	}
	wg.Wait()

}

File: src/1115/main.cc
#include <iostream>
#include <stdlib.h>
#include <vector>
#include <map>
#include <set>
#include <algorithm>
#include <thread>

using namespace std;

// 1  互斥锁

class FooBar {
public:
    FooBar(int n) {
        this->n = n;
        mutex2.lock();
    }

    void foo() {
        
        for (int i = 0; i < n; i++) {
            mutex1.lock();
        	// printFoo() outputs "foo". Do not change or remove this line.
        	printFoo();
            mutex2.unlock();

        }
    }

    void bar() {
        
        for (int i = 0; i < n; i++) {
            mutex2.lock();
        	// printBar() outputs "bar". Do not change or remove this line.
        	printBar();
            mutex1.unlock();
        }
    }

    void printFoo(){
        printf("foo");
    }

     void printBar(){
        printf("bar");
    }


private:
    int n;
    mutex mutex1,mutex2;

};

// 2  信号量


// 3  条件变量


int main() {
    FooBar * obj = new FooBar(5);

    // 开启两个线程 
    thread t1(obj->foo());

    t1.join();


    return 0;
}


 update user set password=password('darwin') where user='root' and host='localhost'; 


